src/qmapshack/realtime/ais/CRtAis.cpp

```diff
@@ -0,0 +1,271 @@
+/**********************************************************************************************
+    Copyright (C) 2018 Oliver Eichler <oliver.eichler@gmx.de>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+**********************************************************************************************/
+
+#include "canvas/CCanvas.h"
+#include "CMainWindow.h"
+#include "helpers/CDraw.h"
+#include "realtime/CRtDraw.h"
+#include "realtime/ais/CRtAis.h"
+#include "realtime/ais/CRtAisInfo.h"
+
+#include <QJsonDocument>
+#include <QtNetwork>
+#include <QtWidgets>
+
+const QString CRtAis::strIcon("://icons/48x48/ActShip.png");
+
+CRtAis::CRtAis(QTreeWidget* parent)
+    : IRtSource(eTypeAis, true, parent)
+{
+    setIcon(eColumnIcon, QIcon(strIcon));
+    setText(eColumnName, "AIS");
+    setCheckState(eColumnCheckBox, Qt::Checked);
+
+    CRtAis::registerWithTreeWidget();
+}
+
+void CRtAis::registerWithTreeWidget()
+{
+    QTreeWidget* tree = treeWidget();
+    if(tree != nullptr)
+    {
+        QTreeWidgetItem* itemInfo = new QTreeWidgetItem(this);
+        itemInfo->setFlags(Qt::ItemIsEnabled | Qt::ItemNeverHasChildren);
+        info = new CRtAisInfo(*this, tree);
+        connect(info, &CRtAisInfo::sigChanged, this, &CRtAis::sigChanged);
+        tree->setItemWidget(itemInfo, eColumnWidget, info);
+        emit sigChanged();
+    }
+}
+
+void CRtAis::loadSettings(QSettings& cfg)
+{
+    QMutexLocker lock(&IRtSource::mutex);
+
+    IRtSource::loadSettings(cfg);
+    showNames = cfg.value("showNames", showNames).toBool();
+
+    if(info != nullptr)
+    {
+        info->loadSettings(cfg);
+    }
+
+    emit sigChanged();
+}
+
+void CRtAis::saveSettings(QSettings& cfg) const
+{
+    QMutexLocker lock(&IRtSource::mutex);
+
+    IRtSource::saveSettings(cfg);
+    cfg.setValue("showNames", showNames);
+
+    if(info != nullptr)
+    {
+        info->saveSettings(cfg);
+    }
+}
+
+QString CRtAis::getDescription() const
+{
+    return tr("<b>AIS</b><br/>"
+              "Get vessel positions via AIS over TCP/IP"
+              );
+}
+
+qint32 CRtAis::getNumberOfShips() const
+{
+    QMutexLocker lock(&IRtSource::mutex);
+    return ships.count();
+}
+
+bool CRtAis::getShowNames() const
+{
+    QMutexLocker lock(&IRtSource::mutex);
+    return showNames;
+}
+
+CRtAis::ship_t* CRtAis::getShipByKey(const QString& key, bool& ok)
```

Comment:
This pointer gives me a shiver. Yes, it's never a nullptr. But returning a `const ship_t&` would express it much better and wouldn't make me scream for a pointer check every time you use it ;)

Reply:
Hehe, got it, I will sort this one out :)